File: lintcode/longest-increasing-subsequence.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/longest-increasing-subsequence/

class Solution:
    """
    @param nums: The integer array
    @return: The length of LIS (longest increasing subsequence)
    """
    def longestIncreasingSubsequence(self, nums):
        # write your code here
        if len(nums) == 0:
            return 0

        dp = [1 for i in xrange(len(nums))]
        res = 0
        for i in xrange(len(nums)):
            for j in xrange(i):
                if nums[j] <= nums[i]:
                    dp[i] = max(dp[j]+1, dp[i])
                    res = max(res, dp[i])
        return res

if __name__ == '__main__':
    s = Solution()
    print s.longestIncreasingSubsequence([4,2,4,5,3,7])
    print s.longestIncreasingSubsequence([1,1,1,1,1,1])
    print s.longestIncreasingSubsequence([5,4,1,2,3])

File: leetcode-2/longest-increasing-subsequence.cc
// Question: https://leetcode.com/problems/longest-increasing-subsequence/

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }

        vector<int> dp(nums.size(), 1);
        for (int i = 0; i < nums.size(); ++i) {
            for (int j = 0; j < i; ++j) {
                if (nums[i] > nums[j]) {
                    dp[i] = max(dp[i], dp[j]+1);
                }
            }
        }

        int ans = INT_MIN;
        for (auto i : dp) {
            ans = max(ans, i);
        }
        return ans;
    }
};
